java/backend.scala

```diff
@@ -70,56 +75,214 @@ class Backend {
+        available = true
+    }
+    
+    if (ss != null) {
+        Try {
+            ss.close();
+        }
+    }
+
+    available
+  }
+  
   def main(args: Array[String]): Unit = {
-    if (args.length < 1) {
-      System.err.println("Usage: Backend <ports-file-output>")
+    if (args.length > 3 || args.length < 2) {
+      System.err.println(
+        "Usage: Backend port id [--service]\n" +
+        "\n" +
+        "  port:      port the gateway will listen to\n" +
+        "  id:        arbitrary numeric identifier for this backend session\n" +
+        "  --service: prevents closing the connection from closing the backend"
+      )
+      
       System.exit(-1)
     }
-
-    val backend = new Backend()
+    
+    port = args(0).toInt
+    sessionId = args(1).toInt
+    isService = args.length > 2 && args(2) == "--service"
+    
+    log("sparklyr session " + sessionId + " ready on port " + port)
+    
     try {
-      // bind to random port
-      val boundPort = backend.init()
-      val serverSocket = new ServerSocket(0, 1, InetAddress.getByName("localhost"))
-      val listenPort = serverSocket.getLocalPort()
-
-      // tell the R process via temporary file
-      val path = args(0)
-      val f = new File(path + ".tmp")
-      val dos = new DataOutputStream(new FileOutputStream(f))
-      dos.writeInt(boundPort)
-      dos.writeInt(listenPort)
-      Serializer.writeString(dos, Utils.rPackages.getOrElse(""))
-      dos.close()
-      f.renameTo(new File(path))
-
-      // wait for the end of stdin, then exit
-      new Thread("wait for socket to close") {
-        setDaemon(true)
-        override def run(): Unit = {
-          // any un-catched exception will also shutdown JVM
-          val buf = new Array[Byte](1024)
-          // shutdown JVM if R does not connect back in 10 seconds
-          serverSocket.setSoTimeout(10000)
-          try {
-            val inSocket = serverSocket.accept()
-            serverSocket.close()
-            // wait for the end of socket, closed if R process die
-            inSocket.getInputStream().read(buf)
-          } finally {
-            backend.close()
-            System.exit(0)
-          }
+      if (portIsAvailable(port))
+      {
+        gatewayServerSocket = new ServerSocket(port, 1, InetAddress.getByName("localhost"))
+      }
+      else
+      {
+        gatewayServerSocket = new ServerSocket(0, 1, InetAddress.getByName("localhost"))
+        val gatewayPort = port
+        port = gatewayServerSocket.getLocalPort()
+        
+        val success = register(gatewayPort, sessionId, port)
+        if (!success) {
+          logError("Failed to register sparklyr session (" + sessionId + ") to gateway port (" + gatewayPort + ")")
+          System.exit(1)
         }
-      }.start()
-
-      backend.run()
+      }
+      
+      gatewayServerSocket.setSoTimeout(0)
+    
+      while(true) {
+        bind()
```

Comment:
@kevinushey Good point! Wrote this instead to describe in details what is going on: https://github.com/rstudio/sparklyr/blob/feature/service-socket/java/backend.scala#L24-L68